allennlp/training/trainer.py

```diff
@@ -297,17 +301,19 @@ def on_batch(
                 self._param_updates[name].sub_(param.detach().cpu())
         else:
             self._param_updates = None
+        # Need to call this before .log_batch(), since .log_batch() updates its
+        # internal batch num counter.
```

Comment:
If this is true, then why does it also take a `batch_number` argument?

Reply:
The `batch_number` argument is the batch number within the current epoch, but `LogWriter` keeps track of the overall batch number.